### PR TITLE
Weighted multisignature wallet integration

### PR DESCRIPTION
**Pull request by gladcow**
Use weighted multisignature wallet for SONs primary wallet

### CODE REVIEW

File: libraries/chain/include/graphene/chain/protocol/son_wallet.hpp [hunk 1]
```diff
@@ -27,6 +27,7 @@ namespace graphene { namespace chain {
         son_wallet_id_type son_wallet_id;
         graphene::peerplays_sidechain::sidechain_type sidechain;
         string address;
+        graphene::peerplays_sidechain::bytes redeem_script;
```
Comment (obucinac) [inline libraries/chain/include/graphene/chain/protocol/son_wallet.hpp]: Redeem sctipt is already stored in address field, as address and a script are tightly paired. This field is not needed and breaks abstraction, by adding bitcoin specific field.

File: libraries/chain/include/graphene/chain/son_wallet_object.hpp [hunk 2]
```diff
@@ -21,6 +21,7 @@ namespace graphene { namespace chain {
          time_point_sec expires;
 
          flat_map<peerplays_sidechain::sidechain_type, string> addresses;
+         flat_map<peerplays_sidechain::sidechain_type, peerplays_sidechain::bytes> redeem_scripts;
```
Comment (obucinac) [inline libraries/chain/include/graphene/chain/son_wallet_object.hpp]: Redeem sctipt is already stored in address field, as address and a script are tightly paired. This field is not needed and breaks abstraction, by adding bitcoin specific field.

File: libraries/chain/son_wallet_evaluator.cpp [hunk 3]
```diff
@@ -72,6 +72,7 @@ object_id_type update_son_wallet_evaluator::do_apply(const son_wallet_update_ope
       if (itr->addresses.find(op.sidechain) == itr->addresses.end()) {
          db().modify(*itr, [&op](son_wallet_object &swo) {
             swo.addresses[op.sidechain] = op.address;
+            swo.redeem_scripts[op.sidechain] = op.redeem_script;
```
Comment (obucinac) [inline libraries/chain/son_wallet_evaluator.cpp]: Redeem sctipt is already stored in address field, as address and a script are tightly paired. This field is not needed and breaks abstraction, by adding bitcoin specific field.

File: libraries/plugins/peerplays_sidechain/include/graphene/peerplays_sidechain/sidechain_net_handler.hpp [hunk 4]
```diff
@@ -32,7 +32,7 @@ class sidechain_net_handler {
     graphene::chain::database& database;
     graphene::peerplays_sidechain::sidechain_type sidechain;
 
-    virtual std::string create_multisignature_wallet( const std::vector<std::string> public_keys ) = 0;
+    virtual std::string create_multisignature_wallet( const std::vector<std::pair<std::string, uint64_t>>& public_keys ) = 0;
```
Comment (obucinac) [inline libraries/plugins/peerplays_sidechain/include/graphene/peerplays_sidechain/sidechain_net_handler.hpp]: Create new method for weighted wallet.

File: libraries/plugins/peerplays_sidechain/peerplays_sidechain_plugin.cpp [hunk 5]
```diff
@@ -499,6 +500,42 @@ void peerplays_sidechain_plugin_impl::on_new_objects(const vector<object_id_type
...
+
+            if(proposal->proposed_transaction.operations.size() == 1
+            && proposal->proposed_transaction.operations[0].which() == chain::operation::tag<chain::bitcoin_transaction_send_operation>::value) {
+               approve_proposal( son_id, proposal->id );
+               continue;
+            }
+         }
+      }
+
+      if( object_id.is<chain::bitcoin_transaction_object>() ) {
```
Comment (obucinac) [inline libraries/plugins/peerplays_sidechain/peerplays_sidechain_plugin.cpp]: All sidechain specific code should be placed in sidechin handler class. We need more abstraction here

File: libraries/chain/include/graphene/chain/protocol/son_wallet.hpp (hunk 1 above)
Comment (gladcow) [reply libraries/chain/include/graphene/chain/protocol/son_wallet.hpp]: What do you mean exactly? Address contains only redeem script hash.

Comment (gladcow) [reply libraries/chain/include/graphene/chain/protocol/son_wallet.hpp]: I see now,  the previous version contains json with all info,  I'll fix it,  thank you

File: libraries/chain/son_wallet_evaluator.cpp (hunk 3 above)
Comment (gladcow) [reply libraries/chain/son_wallet_evaluator.cpp]: ok

File: libraries/chain/include/graphene/chain/son_wallet_object.hpp (hunk 2 above)
Comment (gladcow) [reply libraries/chain/include/graphene/chain/son_wallet_object.hpp]: ok

File: libraries/plugins/peerplays_sidechain/include/graphene/peerplays_sidechain/sidechain_net_handler.hpp (hunk 4 above)
Comment (gladcow) [reply libraries/plugins/peerplays_sidechain/include/graphene/peerplays_sidechain/sidechain_net_handler.hpp]: ok

File: libraries/plugins/peerplays_sidechain/peerplays_sidechain_plugin.cpp (hunk 5 above)
Comment (gladcow) [reply libraries/plugins/peerplays_sidechain/peerplays_sidechain_plugin.cpp]: ok,  I see

File: libraries/chain/include/graphene/chain/protocol/sidechain_transaction.hpp [hunk 6]
```diff
@@ -26,7 +28,7 @@ namespace graphene { namespace chain {
 
       asset                         fee;
       account_id_type               payer;
-      proposal_id_type              proposal_id;
+      bitcoin_transaction_id_type   tx_id;
```
Comment (gladcow) [inline libraries/chain/include/graphene/chain/protocol/sidechain_transaction.hpp]: I'm not sure really is it correct way or not. 
SONs can verify every set of signatures separately or verify them all at the end, and they should verify them at the end in any case,  so I have decided to gather signatures without verification and verify all of them at the end.

Comment (satyakoneru) [reply libraries/chain/include/graphene/chain/protocol/sidechain_transaction.hpp]: If we delink the proposal approval process from the signing procedure, there will be a delay in signature collection. There will be one extra step(solitary approval of send proposal) now with these new changes.
Earlier:
send -> sign by multiple sons and proposal approval in one go -> process transaction object and send out the signed tx.

Now:
send -> proposal approval by multiple sons -> signing by multiple sons -> process transaction object and send out the signed tx.

Comment (gladcow) [reply libraries/chain/include/graphene/chain/protocol/sidechain_transaction.hpp]: Yes,  you are right,  I've reverted it back to proposal already.) I'm not sure it works correctly when inactive SONs should sign the tx (during the primary wallet recreation), I'll try to write some tests for this case.

File: libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp [hunk 7]
```diff
@@ -615,70 +619,152 @@ void sidechain_net_handler_bitcoin::recreate_primary_wallet() {
...
+         auto it = tx_object.signatures.find(son_id);
+         if (it == tx_object.signatures.end())
+            continue;
+         if (it->second.empty())
+         {
+            bitcoin_transaction_sign_operation op;
+            son_object s_obj= plugin.get_son_object(son_id);
+            op.payer = s_obj.son_account;
+            op.proposal_id = proposal.id;
+            fc::ecc::private_key k = plugin.get_private_key(son_id);
```
Comment (sierra19XX) [inline libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp]: If we add signature here, tx_signed for the son rewards is not updated for that son.

Comment (gladcow) [reply libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp]: Yes,  you are right. This code was written before son rewards were merged to the base branch,  and it is possible it is not match the son reward code,  I'll check and fix it.

**Comment by gladcow** [comment]
@sierra19XX ,  I've fixed the issue,  re-review,  please.

File: libraries/plugins/peerplays_sidechain/bitcoin_utils.cpp [hunk 8]
```diff
@@ -677,4 +759,25 @@ bytes add_signatures_to_unsigned_tx(const bytes &unsigned_tx, const std::vector<
...
 }
 
+std::string get_weighted_multisig_address(const std::vector<std::pair<std::string, uint64_t>> &public_keys) {
+   std::vector<std::pair<fc::ecc::public_key, uint64_t>> key_data;
+   for (auto p : public_keys) {
+      fc::ecc::public_key_data kd;
+      for (uint32_t i = 0; i < p.first.length(); i += 2) {
+         std::string byte_str = p.first.substr(i, 2);
+         kd.data[i / 2] = (char)std::strtol(byte_str.c_str(), NULL, 16);
+      }
```
Comment (obucina) [inline libraries/plugins/peerplays_sidechain/bitcoin_utils.cpp]: Bitcoin public keys in configuration file are HEX encoded. I changed this part to read them properly.
However, The final result is invalid address. Eg, for default keys we are using, we get invalid address:
```
03456772301e221026269d3095ab5cb623fc239835b583ae4632f99a15107ef275
02d67c26cf20153fe7625ca1454222d3b3aeb53b122d8a0f7d32a3dd4b2c2016f4
025f7cfda933516fd590c5a34ad4a68e3143b6f4155a64b3aab2c55fb851150f61
0228155bb1ddcd11c7f14a2752565178023aa963f84ea6b6a052bddebad6fe9866
037500441cfb4484da377073459511823b344f1ef0d46bac1efd4c7c466746f666
02ef0d79bfdb99ab0be674b1d5d06c24debd74bffdc28d466633d6668cc281cccf
0317941e4219548682fb8d8e172f0a8ce4d83ce21272435c85d598558c8e060b7f
0266065b27f7e3d3ad45b471b1cd4e02de73fc4737dc2679915a45e293c5adcf84
023821cc3da7be9e8cdceb8f146e9ddd78a9519875ecc5b42fe645af690544bccf
0229ff2b2106b76c27c393e82d71c20eec32bcf1f0cf1a9aca8a237269a67ff3e5
024d113381cc09deb8a6da62e0470644d1a06de82be2725b5052668c8845a4a8da
03df2462a5a2f681a3896f61964a65566ff77448be9a55a6da18506fd9c6c051c1
02bafba3096f546cc5831ce1e49ba7142478a659f2d689bbc70ed37235255172a8
0287bcbd4f5d357f89a86979b386402445d7e9a5dccfd16146d1d2ab0dc2c32ae8
02053859d76aa375d6f343a60e3678e906c008015e32fe4712b1fd2b26473bdd73
Bitcoin RPC call importaddress with body {"jsonrpc": "1.0", "id":"importaddress", "method": "importaddress", "params": ["bc1q6a5l7jw8q9kmvcv3yjfyncp0fk69q0vcwslty5e800xd4ugm8s9qjag62c"] } failed with reply '{"result":null,"error":{"code":-5,"message":"Invalid Bitcoin address or script"},"id":"importaddress"}

The address we get is:
bc1q6a5l7jw8q9kmvcv3yjfyncp0fk69q0vcwslty5e800xd4ugm8s9qjag62c
```

Comment (gladcow) [reply libraries/plugins/peerplays_sidechain/bitcoin_utils.cpp]: It is a mainnet address,  you are trying to publish it in regtest network.Mainnet addresses are started with `bc`,  testnet and regtest addresses are started with `tb`

File: libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp [hunk 9]
```diff
@@ -619,53 +621,47 @@ void sidechain_net_handler_bitcoin::recreate_primary_wallet() {
...
+         op.sidechain = sidechain_type::bitcoin;
+         op.address = address;
+
+         proposal_create_operation proposal_op;
+         proposal_op.fee_paying_account = plugin.get_son_object(plugin.get_current_son_id()).son_account;
+         proposal_op.proposed_ops.emplace_back(op_wrapper(op));
+         uint32_t lifetime = (gpo.parameters.block_interval * gpo.active_witnesses.size()) * 3;
+         proposal_op.expiration_time = time_point_sec(database.head_block_time().sec_since_epoch() + lifetime);
+
+         signed_transaction trx = database.create_signed_transaction(plugin.get_private_key(plugin.get_current_son_id()), proposal_op);
```
Comment (gladcow) [inline libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp]: Is it a correct way to publish tx with `op.payer = GRAPHENE_SON_ACCOUNT;`?

**Comment by gladcow** [comment]
Not actual,  replaced by https://github.com/peerplays-network/peerplays/pull/316

**Pull request closed without merging**